src/components/ElementSchemes/IdScheme.js

```diff
@@ -1,63 +1,80 @@
+import React from 'react'
+import { useDataQuery } from '@dhis2/app-runtime'
 import PropTypes from 'prop-types'
 import i18n from '@dhis2/d2-i18n'
 
-import { fetchAttributes } from '../../utils/helper'
 import { idSchemeOptions } from '../../utils/options'
 import { SelectField } from '../'
 
+const schemeQuery = {
+    org: {
+        resource: 'attributes',
+        params: {
+            filter: ['unique:eq:true', 'organisationUnitAttribute:eq:true'],
+            fields: ['id', 'displayName'],
+            paging: 'false',
+        },
+    },
+    data: {
+        resource: 'attributes',
+        params: {
+            filter: ['unique:eq:true', 'dataElementAttribute:eq:true'],
+            fields: ['id', 'displayName'],
+            paging: 'false',
+        },
+    },
+}
+
 const attributeFoundIn = (attribute, collection) =>
     !!collection.find(({ value }) => value === attribute.value)
 
-const IdScheme = ({ name, label, dataTest }) => {
-    const { baseUrl } = useConfig()
-    const [loading, setLoading] = useState(true)
-    const [schemes, setSchemes] = useState([])
-    const [error, setError] = useState(undefined)
-
-    useEffect(() => {
-        const f = async () => {
-            let err
-
-            const dataElementAttributes = await fetchAttributes(
-                `${baseUrl}/api/`,
-                'dataElementAttribute'
-            ).catch(error => (err = error))
-            const organisationUnitAttributes = await fetchAttributes(
-                `${baseUrl}/api/`,
-                'organisationUnitAttribute'
-            ).catch(error => (err = error))
+const findSharedAttributes = (
+    rawDataElementAttributes,
+    rawOrganisationUnitAttributes
+) => {
+    const dataElementAttributes = rawDataElementAttributes.map(
+        ({ id, displayName }) => ({
+            value: `ATTRIBUTE:${id}`,
+            label: displayName,
+        })
+    )
 
-            setError(err)
+    const organisationUnitAttributes = rawOrganisationUnitAttributes.map(
+        ({ id, displayName }) => ({
+            value: `ATTRIBUTE:${id}`,
+            label: displayName,
+        })
+    )
 
-            if (!err) {
-                const sharedAttributes = dataElementAttributes.reduce(
-                    (shared, attribute) => {
-                        const foundInOrgUnits = attributeFoundIn(
-                            attribute,
-                            organisationUnitAttributes
-                        )
-                        return foundInOrgUnits ? [...shared, attribute] : shared
-                    },
-                    []
-                )
+    const sharedAttributes = dataElementAttributes.reduce(
+        (shared, attribute) => {
+            const foundInOrgUnits = attributeFoundIn(
+                attribute,
+                organisationUnitAttributes
+            )
+            return foundInOrgUnits ? [...shared, attribute] : shared
+        },
+        []
+    )
```

Comment:
Not a biggie, but wouldn't it make more sense to first find the shared attributes and then map to the actual option format? Then it doesn't need to be done individually for both the data element and the org unit attributes

Reply:
Yeah, it makes more sense. This code is copied from the pre-refactor version. Will change.